PR review of Cluster submission implementation

MortenHolmRep opened the pull request:
Initial integration of tom stuttards cluster submission tool.
an example script called train_model_cluster.py has been included in the `graphnet/examples` folder.

running `train_model_cluster.py` executes the `train_model_cluster.sh` script, ensuring that the environment we defined is pulled along into the cluster, which in turn calls the ordinary `train_model.py` script.
We, therefore, do not need to alter the current workflow to submit our scripts to the cluster and only have to deal with the submission parameters.

The current iteration should be able to run on NBI, NPX, DESY and GRID. Although Tom has warned about the extra hurdles of setting it up on GRID. It is however something we may need later down the road.
The submission script automatically detects which cluster you have access to and submits the job to the given cluster. Meaning, that if we have cloned the git repo to hep, it will use the hep cluster for the job.

Any ideas for changes, reductions, etc. are welcome

asogaard commented:
Hi @MortenHolmRep,

Thanks for this extensive PR! :)  It is a lot of material, so I'll try to have a thorough look through. In the meantime:

* Would you mind running `black` on the modified/added files, and/or install the pre-commit hooks to do this automatically? (see https://github.com/icecube/graphnet/blob/main/CONTRIBUTING.md#code-quality).
* I will prepare a PR to resolve the build error in the `lint` job and tag you there.

MortenHolmRep commented:
@asogaard I tried following the instruction in the link, However, after using `pip install -e .[develop]` , I was not able to use the command `pre-commit install` as indicated in the code quality section. It returns the error message `bash: pre-commit: command not found`

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,815 @@
+#     #
+
+#     #Create the submit file this DAGMan file will use
+#     create_condor_submit_file(condor_file_path,export_env=export_env)
+
+#     # Make file
+#     with open(dagman_file_path,"w") as dagman_file :
+
+#         # Write a header
+#         dagman_file.write( "# Autogenerated using fridge/utils/cluster/condor.create_dagman_submit_file\n" )
+
+#         # Loop over jobs provided (each one is a list of args)
+#         for i_job,exe_arg_str in enumerate(exe_args) :
+
+#             # Write job line
+#             job_index_str = "%08i" % i_job
+#             job_id = "job_" + job_index_str
+#             dagman_file.write( "JOB %s %s\n" % (job_id,condor_file_path) )
+
+#             # Define generic args used for all jobs
+#             arg_string = "JobIndex=\"%s\"" % (job_index_str)
+#             arg_string += " Exe=\"%s\"" % (exe_path)
+#             arg_string += " ExeArgs=\"%s\"" % (exe_arg_str)
+#             arg_string += " Memory=\"%i\"" % (memory)
+#             arg_string += " NumGPU=\"%i\"" % (num_gpus)
+#             arg_string += " SubmitDir=\"%s\"" % (submit_dir)
+#             arg_string += " LogDir=\"%s\"" % (log_dir)
+
+#             # Add any bespoke arguments
+#             #TODO This is currently unused and probably needs refurbishing
+#             #for i_arg in range(0,len(arg_names)) :
+#             #    arg_string += " %s=\"%s\" " % (arg_names[i_arg],job.args[i_arg])
+
+#             # Write the args for the job
+#             dagman_file.write( "VARS %s %s\n" % (job_id,arg_string) )
+
+#     #Done
+#     print "DAGMan submit file written : %s" % (dagman_file_path)
+#     return dagman_file_path,condor_file_path
+
```
Remove commented-out lines
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/unix_tools.py:
```diff
@@ -0,0 +1,57 @@
+        raise Exception( "File \"%s\" does not exist" % (filePath) )
+
+
+#which
+#Stolen from: https://stackoverflow.com/questions/377017/test-if-executable-exists-in-python
+def which(program):
+
+    def is_exe(fpath):
+        return os.path.isfile(fpath) and os.access(fpath, os.X_OK)
+
+    fpath, fname = os.path.split(program)
+    if fpath:
+        if is_exe(program):
+            return program
+    else:
+        for path in os.environ["PATH"].split(os.pathsep):
+            path = path.strip('"')
+            exe_file = os.path.join(path, program)
+            if is_exe(exe_file):
+                return exe_file
+
+    return None
```
No used elsewhere?
```suggestion
```

MortenHolmRep replied:
removed unused function

asogaard commented on src/graphnet/utils/cluster/unix_tools.py:
```diff
@@ -0,0 +1,57 @@
+'''
+Tom Stuttard
+'''
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+#Unix specific-tools
+#Tom Stuttard
+
+from future import standard_library
+standard_library.install_aliases()
```
Needed?
```suggestion
```

MortenHolmRep replied:
Tested its dependency and appears to not be needed

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,219 @@
+
+    assert len(file_list) > 0, "No files provided to remove"
+
+    # Let user know what is being removed
+    print("%i files%s to remove :" % (len(file_list), "/directories" if remove_dirs else "") )
+    for i, f in enumerate(file_list) :
+        if i > 50 :
+            print("... (%i more files)" % (len(file_list)-i) )
+            break
+        print(("  %s" % f))
+
+    # Check if user wants to remove these files using a command line prompt
+    if prompt :
+        delete_files = False
+        while True :
+            response = input( "Proceed with deleting these %i files%s? [y/n]\n" % (len(file_list), "/directories" if remove_dirs else "") )
+            if response.lower() == "y" :
+                delete_files = True
+                print("Deleting files...")
+                break
+            elif response.lower() == "n" :
+                print("Not deleting files, skipping...")
+                delete_files = False
+                break
+            else :
+                print("Unrecognised response '%s', choose from [y/n]" % response)
+    else :
+        delete_files = True
+
+    # Remove the output files
+    if delete_files :
+        for f in file_list :
+            if os.path.isfile(f) :
+                os.remove(f)
+            elif os.path.isdir(f) :
+                assert remove_dirs == True, "Found a directory to delete, but `remove_dirs` is False"
+                shutil.rmtree(f)
+
+
+    return delete_files
```
Not used elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,219 @@
+    return [ d for d in glob.glob( dir_path + "/*" ) if not os.path.isdir(d) ]
+
+
+#Test if file locking is supported by the file system / mounting options for a directory
+def is_file_lock_possible(dir_path) :
+
+    import fcntl, tempfile
+
+    #Check directory to test exists
+    if os.path.isdir(dir_path) :
+
+        #Create a temporary file (auto-deleted)
+        with tempfile.NamedTemporaryFile(mode='w+',dir=dir_path) as fd :
+
+            #Try to lock the file
+            #If successful, unlock and return True
+            #Otherwise returns False
+            try :
+                fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                fcntl.flock(fd, fcntl.LOCK_UN)
+                return True
+            except e :
+                return False
+
+    else :
+        raise Exception( "Cannot test locking in '%s', does not exist" % dir_path )
```
Not used elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,219 @@
+    return [ d for d in glob.glob( dir_path + "/*" ) if os.path.isdir(d) ]
+
+
+#Get all files in a directory (NOT recursive)
+def get_files_in_dir(dir_path) :
+    if not os.path.isdir(dir_path) : 
+        raise Exception( "Could not find directory \"%s\", cannot get files in directory" % (dir_path) )
+    return [ d for d in glob.glob( dir_path + "/*" ) if not os.path.isdir(d) ]
```
Not used elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,219 @@
+        raise Exception( "File \"%s\" does not exist, cannot get last modified time" )
+
+
+#Get all subdirectories from a directory
+def get_subdirs(dir_path) :
+    if not os.path.isdir(dir_path) : 
+        raise Exception( "Could not find directory \"%s\", cannot get subdirectories" % (dir_path) )
+    return [ d for d in glob.glob( dir_path + "/*" ) if os.path.isdir(d) ]
```
Not used elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,219 @@
+        numBytes /= 1024.0
+
+
+#Get file last modified time as datetime
+def get_file_mod_time(file_path) :
+    if os.path.exists(file_path) :
+        return datetime.datetime.fromtimestamp(os.path.getmtime(file_path))
+    else :
+        raise Exception( "File \"%s\" does not exist, cannot get last modified time" )
```
Not used elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,219 @@
+        raise Exception( "File \"%s\" does not exist, cannot check file size" )
+
+
+#Get nicely formatted units for file size (from http://stackoverflow.com/questions/2104080/how-to-check-file-size-in-python)
+def format_num_bytes(numBytes) :
+    for unit in ['[bytes]', '[kB]', '[MB]', '[GB]', '[TB]', '[PB]', '[EB]', '[ZB]', '[YB]', '[XB]' ] :
+        if numBytes < 1024.0:
+            return "%0.3g %s" % (numBytes, unit)
+        numBytes /= 1024.0
```
Not used elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/containers.py:
```diff
@@ -0,0 +1,294 @@
+standard_library.install_aliases()
+from builtins import object
+
+#from container_tools import trim_duplicates
+#Return the list with duplicates trimmed (still keeping one of each duplicate), maintaining the order of the original list
+#Note: Cannot just use list(set(l)), as the order of the list is not maintained by the set 
+def trim_duplicates(a_list) :
+    output_list = list()
+    unique_elements = set()
+    for val in a_list :
+        if val not in unique_elements : #e.g. is this value has not already been seen
+            output_list.append(val)
+            unique_elements.add(val)
+    return output_list
```
Not used elsewhere? (Apart from function that are themselves not used elsewhere.)
```suggestion
```

asogaard reviewed (commented):
Hi @MortenHolmRep,
Dropping the first round of suggestions, mainly for removing functions that do not seem to be used elsewhere. Let me know whether I'm completely off base before I proceed. :)

MortenHolmRep commented on examples/Readme.txt:
```diff
@@ -0,0 +1,18 @@
+Contained within this folder are example scripts for running the graphnet framework and how to use them.
```
Now renamed to README.md

MortenHolmRep commented on examples/train_model_cluster.sh:
```diff
@@ -0,0 +1,12 @@
+# conda work environment activated
+source /groups/icecube/${USER}/.bashrc
+cd /groups/icecube/${USER}/graphnet_user/env/
+conda activate gnn_py38_leon
```
This does presuppose that the required conda environment is available and currently that the path exists.
This is currently an issue I encountered on the NPX cluster as .bashrc does not exist, only bash_history.

I am currently in the process of reiterating this script to be more general. I will try and add a submission checker

asogaard replied:
Just bumping this. I think it would be good to have the script which is executed be as self-contained as possible. We might, e.g., want to create a conda environment on the fly, install the necessary packages, and delete the conda environment or something like that, to make sure that we don't have to pre-install the package on the cluster nodes on which we want to run.

asogaard commented:
Hi @MortenHolmRep,
I left some comments here a while back, that seems to have been made "outdated" by black. Should I continue reviewing this, or are there still things you are working on here?

MortenHolmRep commented:
> Hi @MortenHolmRep, I left some comments here a while back, that seems to have been made "outdated" by black. Should I continue reviewing this, or are there still things you are working on here?

I should have gone through your comments regarding the dependencies, converted to black, tested that it still works and removed a few extra things i noticed myself.
So you should be able to review it :)

asogaard commented on examples/train_model_cluster.py:
```diff
@@ -0,0 +1,34 @@
+"""
+
+import os
+from graphnet.utils.cluster_params import *
```
I would rename `src/graphnet/utils` -> `src/graphnet/utilities` (since the latter already exists), and in `src/graphnet/utilities/__init__.py` add the line
```
from .cluster_submitter import ClusterSubmitter
```
(provided `cluster_params.py` is renamed to `cluster_submitter.py`) such that the below syntax becomes possible
```suggestion
from graphnet.utilities.cluster import ClusterSubmitter
```

MortenHolmRep replied:
I agree. This is a more elegant solution along with the directory cleaning.

asogaard commented on examples/train_model.py:
```diff
@@ -26,7 +26,7 @@
 
 # Constants
 features = FEATURES.DEEPCORE
```
Given that this file does not contain any substantial changes that relate to the topic of this PR, I would revert its changes (such that it doesn't show up under "Changed files").

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+standard_library.install_aliases()
+import os
+
+from graphnet.utils.cluster.filesys_tools import make_dir, is_executable
+from graphnet.utils.cluster.unix_tools import tail
+
+from graphnet.utils.cluster.job import JOB_INDEX_FMT, JOB_NAME_FMT
```
```suggestion
from graphnet.utilities.cluster.filesys_tools import make_dir, is_executable
from graphnet.utilities.cluster.unix_tools import tail

from graphnet.utilities.cluster.job import JOB_INDEX_FMT, JOB_NAME_FMT
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+
+
+def create_condor_submit_file(
+    file_path,
+    memory_MB=1000,
+    disk_space_MB=1000,
+    num_cpus=1,
+    num_gpus=0,
+    export_env=False,
+    require_cvmfs=False,
+    require_avx=False,
+    require_cuda=False,
+    require_sl7=False,  # Scientific Linux 7 (+ variants)
+    choose_sites=None,
+    exclude_sites=None,
+    user_proxy=False,
+    wall_time_hr=None,
+    accounting_group=None,
```
Type hints?

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+
+# Create a DAGMan submit file
+def create_dagman_submit_file(
+    submit_dir,
+    log_dir,
+    dagman_file_name,
+    condor_file_name,
+    jobs,
+    **submit_file_kw,
```
Type hints?

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+        return None, None, None
+
+
+#
+# Test
+#
+
+if __name__ == "__main__":
+
+    from .job import ClusterCommand, ClusterJob
+
+    test_dir = "./tmp/condor"
+    make_dir(test_dir)
+
+    exe_path = "echo"
+    exe_args = [
+        "bar",
+        "bar",
+    ]
+
+    dagman_file, condor_file = create_dagman_submit_file(
+        submit_dir=test_dir,
+        log_dir=test_dir,
+        dagman_file_name="test.dagman",
+        condor_file_name="test.condor",
+        exe_path=exe_path,
+        exe_args=exe_args,
+        memory=1000,
+    )
```
Move to a unit test?

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+            )
+            returnString += "    Execution duration : %s\n" % (
+                self.logFileInfo.getExecuteDuration()
+                if self.logFileInfo.getExecuteDuration() != None
+                else "[Not parsed]"
+            )
+            returnString += "    Memory used        : %s\n" % (
+                "%i [MB]" % self.logFileInfo.memoryUsedMB
+                if self.logFileInfo.memoryUsedMB != None
+                else "[Not parsed]"
+            )
+            returnString += "    Memory requested   : %s\n" % (
+                "%i [MB]" % self.logFileInfo.memoryRequestedMB
+                if self.logFileInfo.memoryRequestedMB != None
+                else "[Not parsed]"
+            )
+            returnString += "    Memory allocated   : %s\n" % (
+                "%i [MB]" % self.logFileInfo.memoryAllocatedMB
+                if self.logFileInfo.memoryAllocatedMB != None
+                else "[Not parsed]"
+            )
+            returnString += "    Return code        : %s\n" % (
+                str(self.logFileInfo.returnCode)
+                if self.logFileInfo.returnCode != None
+                else "[Not parsed]"
+            )
+        else:
+            returnString += "  [Log file not parsed]\n"
+
+        if self.outFileInfo != None:
+            returnString += "  Out file info :\n"
+            returnString += "    Num sub jobs       : %s\n" % (
+                self.outFileInfo.numCommands
+                if self.outFileInfo.numCommands != None
+                else "[Not parsed]"
+            )
+        else:
+            returnString += "  [Out file not parsed]\n"
+
+        return returnString
```
I don't see this class referenced elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+        return returnString
+
+
+# Subclass to be filled with parsed out file
+class CondorJobMetricsOutFileInfo(object):
+    def __init__(self, outFilePath):
+        self.numCommands = None
+        self.parse(outFilePath)
+
+    def fullyParsed(self):
+        return self.numCommands != None
+
+    def parse(self, outFilePath):
+
+        # Check file exists, open it, and read into a string
+        if os.path.exists(outFilePath):
+            self.outFilePath = outFilePath
+            with open(self.outFilePath) as outFile:
+                outFileText = outFile.read()
+
+                # Parse num commands
+                self.numCommands = self.getNumCommandsFromFile(outFileText)
+
+    def getNumCommandsFromFile(self, outFileText):
+
+        # Split text into lines
+        lines = outFileText.split("\n")
+
+        # Find num commands line
+        key = "JOB INFO : Num commands = "
+        for line in lines:
+            keyPos = line.find(key)
+            if keyPos > -1:
+
+                # Num commands is rest of line after the key
+                return int(line[keyPos + len(key) :])
+
+        # If here, didn't manage to parse
+        return None
```
I don't see this class referenced elsewhere (apart from one or more places that are not themselves references elsewhere)?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+        returnCodeLength = 1
+        key = "return value "
+        keyPos = logFileText.find(key)
+        if keyPos > -1:  # check found key
+            returnCodePos = keyPos + len(key)
+            if len(logFileText) > returnCodePos + 1:
+
+                # Grad the return code, handling -ve values
+                returnCode = logFileText[
+                    returnCodePos : returnCodePos + returnCodeLength
+                ]
+                if returnCode == "-":
+                    returnCode += logFileText[
+                        returnCodePos
+                        + returnCodeLength : returnCodePos
+                        + returnCodeLength
+                        + 1
+                    ]
+                return int(returnCode)
+
+        # If here, didn't manage to parse
+        return None
+
+    def getMemoryUsageFromFile(self, logFileText):
+
+        # Parse memory usage from file
+        # Break into lines, find line with table containing memory info, then tokenise the oine and get the cells containing the memory used and requested
+        key = "Memory (MB)"
+        logFileLines = logFileText.split("\n")
+        for line in logFileLines:
+            if key in line:
+                tokens = line.split()
+                if len(tokens) == 6:
+                    memoryUsed = int(tokens[3])
+                    memoryRequested = int(tokens[4])
+                    memoryAllocated = int(tokens[5])
+                    return memoryUsed, memoryRequested, memoryAllocated
+
+        # If here, didn't manage to parse
+        return None, None, None
```
I don't see this class referenced elsewhere (apart from one or more places that are not themselves references elsewhere)?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+
+            # Parse the return status code
+            self.returnCode = self.getReturnCode(dagmanOutputFilePath)
+
+        else:
+            raise Exception(
+                'DAGMan .out file "%s" does not exist' % dagmanOutputFilePath
+            )
+
+    def fullyParsed(self):
+        return self.returnCode != None
+
+    def getReturnCode(self, dagmanOutputFilePath):
+
+        # Get the last line of the file (which has the status, assuming it has completed)
+        statusLines = tail(dagmanOutputFilePath, 1)
+
+        # Check got the one line I expected
+        if len(statusLines) == 1:
+
+            statusLine = statusLines[0]
+
+            # Parse status from this line
+            statusKey = "EXITING WITH STATUS"
+            if statusKey in statusLine:
+
+                # Extract the status key and return it
+                statusKeyStartIndex = statusLine.find(statusKey)
+                statusCodeStartIndex = (
+                    statusKeyStartIndex + len(statusKey) + 1
+                )  # +1 for the space after it
+                if len(statusLine) > (statusCodeStartIndex + 1):
+                    return int(
+                        statusLine[
+                            statusCodeStartIndex : statusCodeStartIndex + 1
+                        ]
+                    )
+
+        # If here, failed in parsing return code (maybe DAG is still running, or died in an uncontrolled way)
+        return None
```
I don't see this class referenced elsewhere (apart from one or more places that are not themselves references elsewhere)?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+            )
+
+
+# Subclass containing metrics file information
+class DAGManMetricsMetricsFileInfo(object):
+    def __init__(self, metricsJsonFilePath):
+        self.parse(metricsJsonFilePath)
+
+    def fullyParsed(self):
+        return True  # No conditional parsing for this file at this time
+
+    def parse(self, metricsJsonFilePath):
+
+        # Open the JSON metrics file and parse using a json parser
+        if os.path.exists(metricsJsonFilePath):
+            with open(metricsJsonFilePath) as dagmanMetricsFile:
+
+                metricsNode = json.load(dagmanMetricsFile)
+                if metricsNode:
+
+                    # Read individual details from the metrics file
+                    self.numJobs = metricsNode["jobs"]
+                    self.numJobsSucceeded = metricsNode["jobs_succeeded"]
+                    self.numJobsFailed = metricsNode["jobs_failed"]
+                    self.start_time = metricsNode["start_time"]
+                    self.end_time = metricsNode["end_time"]
+                    self.duration = metricsNode["duration"]
+                    # self.jobsCombinedDuration = metricsNode['total_job_time'] #TODO replace once DAGman actually writes this variable, currently always 0
+                    self.dag_status = metricsNode["dag_status"]
+
+                else:
+                    raise Exception(
+                        'DAGMan metrics file "%s" cannot be parsed as a JSON file'
+                        % metricsJsonFilePath
+                    )
+
+        else:
+            raise Exception(
+                'DAGMan metrics file "%s" does not exist' % metricsJsonFilePath
+            )
```
I don't see this class referenced elsewhere (apart from one or more places that are not themselves references elsewhere)?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+        return returnString
+
+
+# Subclass containing submit file information
+class DAGManMetricsSubmitFileInfo(object):
+    def __init__(self, dagmanSubmitFilePath):
+        self.parse(dagmanSubmitFilePath)
+
+    def fullyParsed(self):
+        return True  # No conditional parsing for this file at this time
+
+    def parse(self, dagmanSubmitFilePath):
+
+        # Check submit exists, open it, and read it into a string
+        if os.path.exists(dagmanSubmitFilePath):
+            with open(dagmanSubmitFilePath) as submit_file:
+                submit_fileText = submit_file.read()
+
+                # Count the instances of the "JOB" ketword in the string
+                self.numJobs = submit_fileText.count("JOB")
+
+        else:
+            raise Exception(
+                'DAGMan submit file "%s" does not exist' % dagmanSubmitFilePath
+            )
```
I don't see this class referenced elsewhere (apart from one or more places that are not themselves references elsewhere)?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+                else "[Not parsed]"
+            )
+        else:
+            returnString += "  [Output file not parsed]\n"
+
+        if self.metricsFileInfo != None:
+            returnString += "  Metrics file info :\n"
+            returnString += "    Num jobs             : %i\n" % (
+                self.metricsFileInfo.numJobs
+            )
+            returnString += "    Num jobs succeeded   : %i (%0.3f%%)\n" % (
+                self.metricsFileInfo.numJobsSucceeded,
+                percentage(
+                    self.metricsFileInfo.numJobsSucceeded,
+                    self.metricsFileInfo.numJobs,
+                ),
+            )
+            returnString += "    Num jobs failed      : %i (%0.3f%%)\n" % (
+                self.metricsFileInfo.numJobsFailed,
+                percentage(
+                    self.metricsFileInfo.numJobsFailed,
+                    self.metricsFileInfo.numJobs,
+                ),
+            )
+            returnString += "    Start time           : %s\n" % (
+                self.metricsFileInfo.start_time
+            )
+            returnString += "    End time             : %s\n" % (
+                self.metricsFileInfo.end_time
+            )
+            returnString += "    Time taken in total  : %s\n" % (
+                time.human_time(self.metricsFileInfo.duration)
+            )
+            returnString += "    DAG status           : %s\n" % (
+                interpretDAGStatus(self.metricsFileInfo.dag_status)
+            )
+        else:
+            returnString += "  [Metrics file not parsed]\n"
+
+        return returnString
```
I don't see this class referenced elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+    return dagman_file_path, condor_file_path
+
+
+#
+# Parsing metrics
+#
+
+# Interpret the DAG status codes (from https://research.cs.wisc.edu/htcondor/manual/latest/2_10DAGMan_Applications.html)
+def interpret_dag_status(dag_status):
+    if dag_status == 0:
+        return "OK"
+    elif dag_status == 1:
+        return "Other error"
+    elif dag_status == 2:
+        return "One or more nodes failed"
+    elif dag_status == 3:
+        return "DAG aborted (ABORT-DAG-ON)"
+    elif dag_status == 4:
+        return "DAG removed (condor_rm)"
+    elif dag_status == 5:
+        return "DAG cycle found"
+    elif dag_status == 6:
+        return "DAG halted"
+    else:
+        raise Exception("Unknown DAG status : %i" % (dag_status))
+
```
I don't see this function referenced elsewhere?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+#
+
+
+def percentage(numerator, denominator):
+    return 100.0 * fraction(numerator, denominator)
+
+
```
I don't see this function referenced elsewhere (apart from one or more places that are not themselves references elsewhere)?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/condor.py:
```diff
@@ -0,0 +1,843 @@
+
+Tom Stuttard
+"""
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import unicode_literals
+
+from builtins import open
+from builtins import int
+from builtins import str
+from builtins import object
+from future import standard_library
+
+standard_library.install_aliases()
```
Are these imports statements strictly necessary?
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,49 @@
+
+# from graphnet.utils.cluster import containers
+
+# py2 vs 3 compatibility
+try:
+    input = raw_input
+except NameError:
+    pass
+
```
We don't support `python2`.
```suggestion
```

asogaard commented on src/graphnet/utils/cluster/filesys_tools.py:
```diff
@@ -0,0 +1,49 @@
+# Tools for handling files and directories
+#
+
+# Create a directory with robust error handling, including race conditions
+def make_dir(dir_path, raise_exception=False):
+    if not os.path.exists(dir_path):
+        try:
+            os.makedirs(dir_path)
+            return True
+        except OSError:
+            assert os.path.isdir(
+                dir_path
+            ), 'Could not create directory "%s"' % (dir_path)
+    return False
```
This seems like a lot of work, if `os.makedirs(path, exists_ok=True)` could be used, but it seems like due consideration was given to this, so I am fine to keep this if it is necessary (as it seems).

asogaard reviewed (commented):
Hi @MortenHolmRep,
Dropping off a fresh batch of comments here, mainly meant to slim down the code we will end up taking responsibility for. :)

MortenHolmRep commented on src/graphnet/utils/cluster_params.py:
```diff
@@ -0,0 +1,913 @@
+"""
```
I agree, the directory structe has been cleaned

RasmusOrsoe commented:
@MortenHolmRep @asogaard  is this still in progress?

MortenHolmRep commented:
@RasmusOrsoe This is currently on hold on my side. currently performing data transformation and then I will look at it again.

asogaard commented:
Hi @MortenHolmRep,
I have labelled this PR as "stale" as it hasn't seen updates in a while and since the feature branch is quite behind `main`. That is not to say that we should necessarily abandon this PR, nor the issue it is trying to solve. It is simple to highlight in the PR overview that, currently, this PR is not being developed on, and to suggest that it might be easier to start a fresh PR off of `main` rather than to continue working on this.
